core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1741,7 +1741,7 @@ private static int fieldCnt(MutableRel rel) {
   }
 
   /** Explain filtering condition and projections from MutableCalc. */
-  private static Pair<RexNode, List<RexNode>> explainCalc(MutableCalc calc) {
+  public static Pair<RexNode, List<RexNode>> explainCalc(MutableCalc calc) {
     final RexShuttle shuttle = getExpandShuttle(calc.program);
     final RexNode condition = shuttle.apply(calc.program.getCondition());
```

Comment:
Do we have to change the modifier?

Reply:
This modifier is used in `org.apache.calcite.test.MaterializationTest.CustomizedMaterializationRule`.
If external rules need to use this method, we should change it.